src/org/javarosa/xpath/expr/XPathPathExpr.java

```diff
@@ -70,209 +69,134 @@
-                }
+            if (step.predicates.length > 0) {
+                List<XPathExpression> v = new ArrayList<XPathExpression>(step.predicates.length);
+                Collections.addAll(v, step.predicates);
                 ref.addPredicate(i, v);
             }
         }
         return ref;
     }
 
-    public XPathNodeset eval (DataInstance m, EvaluationContext ec) {
-        TreeReference genericRef = getReference();
-
-        TreeReference ref;
-        if(genericRef.getContext() == TreeReference.CONTEXT_ORIGINAL) {
-            ref = genericRef.contextualize(ec.getOriginalContext());
-        } else {
-            ref = genericRef.contextualize(ec.getContextRef());
-        }
-
-        //We don't necessarily know the model we want to be working with until we've contextualized the
-        //node
-
-        //check if this nodeset refers to a non-main instance
-        if(ref.getInstanceName() != null && ref.isAbsolute())
-        {
-            DataInstance nonMain = ec.getInstance(ref.getInstanceName());
-            if(nonMain != null)
-            {
-                m = nonMain;
-            }
-            else
-            {
-                throw new XPathMissingInstanceException(ref.getInstanceName(), "Instance referenced by " + ref.toString(true) + " does not exist");
-            }
-        } else {
-            //TODO: We should really stop passing 'm' around and start just getting the right instance from ec
-            //at a more central level
-            m = ec.getMainInstance();
-
-            if(m == null) {
-                    String refStr = ref == null ? "" : ref.toString(true);
-                    throw new XPathException("Cannot evaluate the reference [" + refStr + "] in the current evaluation context. No default instance has been declared!");
-            }
-        }
-
-        // regardless of the above, we want to ensure there is a definition
-        if(m.getRoot() == null) {
-            //This instance is _declared_, but doesn't actually have any data in it.
-            throw new XPathMissingInstanceException(ref.getInstanceName(), "Instance referenced by " + ref.toString(true) + " has not been loaded");
-        }
-
-        // this makes no sense...
-//        if (ref.isAbsolute() && m.getTemplatePath(ref) == null) {
-//            List<TreeReference> nodesetRefs = new List<TreeReference>();
-//            return new XPathNodeset(nodesetRefs, m, ec);
-//        }
-
-      List<TreeReference> nodesetRefs = ec.expandReference(ref);
-
-        //to fix conditions based on non-relevant data, filter the nodeset by relevancy
-        for (int i = 0; i < nodesetRefs.size(); i++) {
-            if (!m.resolveReference(nodesetRefs.get(i)).isRelevant()) {
-                nodesetRefs.remove(i);
-                i--;
-            }
-        }
-
-        return new XPathNodeset(nodesetRefs, m, ec);
+    public XPathNodeset eval(final DataInstance unusedDataInstance, final EvaluationContext ec) {
```

Comment:
I would argue to remove these `final`s here. Maybe the one for `TreeReference ref` in `getReference()` is useful while it gets smaller but here it's just one line of code and those variables are just passed downstream.

Reply:
Thanks. I intended to remove them, and will now. They were part of the evolution of the code (originally the first parameter was never read, but reassigned!).

Reply:
Wow, it's great that you fixed that one :)

Reply:
At least I made it obvious that all those callers are passing in something that is ignored.